core/src/main/java/org/apache/iceberg/variants/PrimitiveWrapper.java

```diff
@@ -0,0 +1,206 @@
+  private static final byte NULL_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_NULL);
+  private static final byte TRUE_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_TRUE);
+  private static final byte FALSE_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_FALSE);
+  private static final byte INT8_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_INT8);
+  private static final byte INT16_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_INT16);
+  private static final byte INT32_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_INT32);
+  private static final byte INT64_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_INT64);
+  private static final byte FLOAT_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_FLOAT);
+  private static final byte DOUBLE_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_DOUBLE);
+  private static final byte DATE_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_DATE);
+  private static final byte TIMESTAMPTZ_HEADER =
+      VariantUtil.primitiveHeader(Primitives.TYPE_TIMESTAMPTZ);
+  private static final byte TIMESTAMPNTZ_HEADER =
+      VariantUtil.primitiveHeader(Primitives.TYPE_TIMESTAMPNTZ);
+  private static final byte DECIMAL4_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_DECIMAL4);
+  private static final byte DECIMAL8_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_DECIMAL8);
+  private static final byte DECIMAL16_HEADER =
+      VariantUtil.primitiveHeader(Primitives.TYPE_DECIMAL16);
+  private static final byte BINARY_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_BINARY);
+  private static final byte STRING_HEADER = VariantUtil.primitiveHeader(Primitives.TYPE_STRING);
+
+  private final Variants.PhysicalType type;
+  private final T value;
+  private ByteBuffer buffer = null;
+
+  PrimitiveWrapper(Variants.PhysicalType type, T value) {
+    this.type = type;
+    this.value = value;
+  }
+
+  @Override
+  public Variants.PhysicalType type() {
+    return type;
+  }
+
+  @Override
+  public T get() {
+    return value;
+  }
+
+  @Override
+  public int sizeInBytes() {
+    switch (type()) {
+      case NULL:
+      case BOOLEAN_TRUE:
+      case BOOLEAN_FALSE:
+        return 1; // 1 header only
+      case INT8:
+        return 2; // 1 header + 1 value
+      case INT16:
+        return 3; // 1 header + 2 value
+      case INT32:
+      case DATE:
+      case FLOAT:
+        return 5; // 1 header + 4 value
+      case INT64:
+      case DOUBLE:
+      case TIMESTAMPTZ:
+      case TIMESTAMPNTZ:
+        return 9; // 1 header + 8 value
+      case DECIMAL4:
+        return 6; // 1 header + 1 scale + 4 unscaled value
+      case DECIMAL8:
+        return 10; // 1 header + 1 scale + 8 unscaled value
+      case DECIMAL16:
+        return 18; // 1 header + 1 scale + 16 unscaled value
+      case BINARY:
+        return 5 + ((ByteBuffer) value).remaining(); // 1 header + 4 length + value length
```

Comment:
This uses the `ByteBuffer` to pass a section of the backing array. If the position or limit of that buffer were changed, it would change the value. You're right that reading part of the buffer could potentially change the position and have that effect, but this is still a valid use of `ByteBuffer`. We use the position and limit to track a section of the storage, not for state.

Reply:
Let's say we have a PrimitiveWrapper of a binary `b = new PrimitiveWrapper( binary, 0x1234)`, would we always expect b.sizeInBytes() always to be 5 + 2 and we shouldn't change that for this method contract?

Reply:
I'm not sure I follow what you're suggesting.

Reply:
Here is what I mean: if I wrap an integer in TRUE value in PrimitiveWrapper, then the size is always 1 (that seems to make sense even if the caller reads the value through PrimitiveWrapper.get()) . While if I wrap a binary, then the size would change if the value is read through PrimitiveWrapper.get().

I may not get "We use the position and limit to track a section of the storage, not for state.". I assume that PrimitiveWrapper accepts a value, then the size would never change?

Reply:
Here is what I mean with the following test: 

primitive input: Variants.of(ByteBuffer.wrap(new byte[] {0x0a, 0x0b, 0x0c, 0x0d}))

@ParameterizedTest
  @FieldSource("PRIMITIVES")
  public void testPrimitiveValueSerialization(VariantPrimitive<?> primitive) {
       ((ByteBuffer)primitive.get()).get();   -- add this line then 

    // write the value to the middle of a large buffer
    int size = primitive.sizeInBytes();
    ByteBuffer buffer = ByteBuffer.allocate(size + 1000).order(ByteOrder.LITTLE_ENDIAN);
    primitive.writeTo(buffer, 300);
...
}

We would still expect to write the whole ByteBuffer of 9 bytes while now we are writing 8 bytes. That is not expected, right?

Reply:
I see. We assume the buffer will not be modified.